File: zZ/java/add-and-search-word-data-structure-design.java
// 211. Add and Search Word - Data structure design
// Difficulty: Medium

// Design a data structure that supports the following two operations:

// void addWord(word)
// bool search(word)
// search(word) can search a literal word or a regular expression string containing only letters a-z or .
// . means it can represent any one letter.

// Your WordDictionary object will be instantiated and called as such:
// WordDictionary wordDictionary;
// wordDictionary.addWord("word");
// wordDictionary.search("pattern");

// For example:

// addWord("bad")
// addWord("dad")
// addWord("mad")
// search("pad") -> false
// search("bad") -> true
// search(".ad") -> true
// search("b..") -> true
// Note:
// You may assume that all words are consist of lowercase letters a-z.

// Hint:
// You should be familiar with how a Trie works. If not, please work on this problem: Implement Trie (Prefix Tree) first.

// Time:  O(min(n, h)), per operation
// Space: O(min(n, h))

public class WordDictionary {
    private class TrieNode {
        boolean isWord;
        char c;
        Map<Character, TrieNode> hm;
        TrieNode() {
            isWord = false;
            leaves = new HashMap<>();
        }
    }
    
    // Adds a word into the data structure.
    private TrieNode root;
    public WordDictionary() {
        root = new TrieNode();
    }
    public void addWord(String word) {
        TrieNode cur = root;
        for (int i = 0; i < word.length(); ++i) {
            if (!cur.leaves.containsKey(word.charAt(i))) {
                TrieNode node = new TrieNode();
                node.c = word.charAt(i);
                cur.leaves.put(word.charAt(i), node);
            }
            cur = cur.leaves.get(word.charAt(i));
        }
        cur.isWord = true;
    }

    // Returns if the word is in the data structure. A word could
    // contain the dot character '.' to represent any one letter.
    public boolean search(String word) {
        return search(word, root, 0);
    }
    
    public boolean search(String word, TrieNode node, int index) {        
        if (index == word.length()) {
            return node.isWord;
        }
        if (node.leaves.containsKey(word.charAt(index))) {
            return search(word, node.leaves.get(word.charAt(index)), index + 1);
        } else if (word.charAt(index) == '.') {
            for(TrieNode leaf : node.leaves.values()) {
                if (search(word, leaf, index + 1)) {
                    return true;
                }
            }
        }
        return false;
    }
}

// Your WordDictionary object will be instantiated and called as such:
// WordDictionary wordDictionary = new WordDictionary();
// wordDictionary.addWord("word");
// wordDictionary.search("pattern");

File: data-structure-and-algorithm/javascript/leetcode/add-and-search-word-data-structure-design.js
/**
Design a data structure that supports adding new words and finding if a string matches any previously added string.

Implement the WordDictionary class:

WordDictionary() Initializes the object.
void addWord(word) Adds word to the data structure, it can be matched later.
bool search(word) Returns true if there is any string in the data structure that matches word or false otherwise. word may contain dots '.' where dots can be matched with any letter.
 

Example:

Input
["WordDictionary","addWord","addWord","addWord","search","search","search","search"]
[[],["bad"],["dad"],["mad"],["pad"],["bad"],[".ad"],["b.."]]
Output
[null,null,null,null,false,true,true,true]

Explanation
WordDictionary wordDictionary = new WordDictionary();
wordDictionary.addWord("bad");
wordDictionary.addWord("dad");
wordDictionary.addWord("mad");
wordDictionary.search("pad"); // return False
wordDictionary.search("bad"); // return True
wordDictionary.search(".ad"); // return True
wordDictionary.search("b.."); // return True
 

Constraints:

1 <= word.length <= 500
word in addWord consists lower-case English letters.
word in search consist of  '.' or lower-case English letters.
At most 50000 calls will be made to addWord and search.
*/

class WordDictionary {
    constructor() {
        this.root = new TrieNode();
    }
    
    /**
     * 
     * @param { string } word 
     */
    addWord(word) {
        let curr = root;
        for (const char of word) {
            if (curr.leaves.hasOwnProperty(char)) {
                curr.leaves[char] = new TrieNode(); 
            }
            curr = curr.leaves[char];
        }
        curr.isWord = true;
    }
    
    /**
     * 
     * @param { string } word 
     * @returns 
     */
    search(word) {
        return this.searchWord(word, this.root, 0);
    }

    /**
     * 
     * @param { string } word
     * @param { TrieNode } node
     * @param { number } index
     * @returns { boolean }
     */
    searchWord(word, node, index) {
        if (index === word.length) {
            return node.isWord;
        }

        const char = word[index];
        if (char !== '.') {
            if (!node.hasOwnProperty(char)) {
                return false;
            }
            return this.searchWord(word, node.leaves[char], index + 1);
        }

        for (const leaf of Object.keys(node.leaves)) {
            if (this.searchWord(word, leaf, index + 1)) {
                return true;
            }
        }

        return false;
    }
}

class TrieNode {
    constructor() {
        this.leaves = {}
        this.isWord = false
    }   
}